Classes/Routing/RoutingService.php

```diff
@@ -0,0 +1,759 @@
+
+        return $this->fetchEnhancerInSiteConfigurationByPageUid(
+            $site,
+            $pageUid
+        );
+    }
+
+    /**
+     * Returns the route enhancer configuration by given site and page uid
+     *
+     * @param Site $site
+     * @param int $pageUid
+     * @return array
+     */
+    public function fetchEnhancerInSiteConfigurationByPageUid(Site $site, int $pageUid): array
+    {
+        $configuration = $site->getConfiguration();
+        if (empty($configuration['routeEnhancers']) || !is_array($configuration['routeEnhancers'])) {
+            return [];
+        }
+        $result = [];
+        foreach ($configuration['routeEnhancers'] as $routing => $settings) {
+            // Not the page we are looking for
+            if (!in_array($pageUid, $settings['limitToPages'])) {
+                continue;
+            }
+            // TODO: Instead of checking a string, check an interface (special interface for combined enhancer)
+            //       This have be enabled by configuration to avoid long rendering times
+            if (empty($settings) || !isset($settings['type']) || $settings['type'] !== 'CombinedFacetEnhancer') {
+                continue;
+            }
+            $result[] = $settings;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Retrieve the site configuration by URI
+     *
+     * @param UriInterface $uri
+     * @return Site|null
+     */
+    public function findSiteByUri(UriInterface $uri): ?Site
+    {
+        $sites = $this->getSiteFinder()->getAllSites();
+        if (count($sites) === 1) {
+            return array_values($sites)[0];
+        }
+
+        foreach ($sites as $siteKey => $site) {
+            if ($site->getBase()->getHost() !== $uri->getHost()) {
+                continue;
+            }
+
+            return $site;
+        }
+
+        return null;
+    }
+
+    /**
+     * Retrieve the site by given UID
+     *
+     * @param int $pageUid
+     * @return Site|null
+     */
+    public function findSiteByUid(int $pageUid): ?Site
+    {
+        try {
+            $site = $this->getSiteFinder()
+                ->getSiteByPageId($pageUid);
+            return $site;
+        } catch (SiteNotFoundException $exception) {
+            return null;
+        }
+    }
+
+    /**
+     * Convert the base string into a URI object
+     *
+     * @param string $base
+     * @return UriInterface|null
+     */
+    public function convertStringIntoUri(string $base): ?UriInterface
+    {
+        try {
+            /* @var Uri $uri */
+            $uri = GeneralUtility::makeInstance(
+                Uri::class,
+                $base
+            );
+
+            return $uri;
+        } catch (\InvalidArgumentException $argumentException) {
+            return null;
+        }
+    }
+
+    /**
+     * Returns the current language
+     * @TODO Improvement required: Currently we expect that the longest length for base is at the end of the language array
```

Comment:
you might want to check out "SiteMatcher" from Core, which should do a similar thing. Maybe we can improve the code in there so we don't need code duplication?

Reply:
I'll check it.

Reply:
I agree the method should replaced with class SiteMatcher. This requires changes how process the request and retrieve the language within the middleware.

Reply:
The own implementation replaced with the TYPO3 own site matcher.